### PR TITLE
Add equivalent to .net Task.Result

### PR DESCRIPTION
**Pull request by stefanrusek**
I would love to replace getResult() rather than add waitAndGetResult(),
since the current getResult() just feels broken; however that would be a
rather big breaking change.

### CODE REVIEW

**Comment by grantland** [comment]
Hrmm this one is tricky since we've diverged a bit from the original implementation here and you're right that updating `getResult()` would be a rather big breaking change.

You're current implementation doesn't really match any existing patterns as it throws unspecified `RuntimeException`s and generally seems unsafe to call without understanding the implementation details. 

A `waitForResult()` that doesn't throw, similar to `waitForCompletion()`, would match the existing APIs a bit better, but would also just be a shortcut to to `waitForCompletion()` -> `getResult()`. Do you think that would still be worthwhile?

/cc @nlutsenko @richardjrossiii @hallucinogen

**Comment by stefanrusek** [comment]
@grantland I agree the exception handling here is less than ideal. my original version did not eat+rethrow the wait error, but then I had to add a _throws_ clause to the method signature, which wasn't really something I was happy with. 

Given that all our existing code that calls waitForCompletion() before calling getResult() already has to deal with this throws clause, so it isn't the biggest deal. I can change it back. I'm just not a huge fan of java's checked exceptions. I can see the logic, but on large projects it just tends to be more pain than benefit.

**Comment by stefanrusek** [comment]
I've added throw clauses for InterruptedException and TimeOutException. One advantage of renaming getResult() to getResultOrNull() and renaming waitAndGetResult() to getResult() now is that the new exceptions will cause compilation errors rather than runtime changes.

**Comment by stefanrusek** [comment]
@grantland I forget to respond to your last question, where you asked if a version of waitForResult() that didn't throw would be useful.

I've used the TPS tons, and now bolts quite a bit. In my opinion, the big advantage of the TPS' Task.Result is that it throws. It allows usages where something gives you a Task. You ask it for a result and you know that either you'll get a result or exception will be interrupted. This allows for error handling where it makes sense rather than the place where you are asking for the result, where a bunch of if statements to check for successful non-cancelled task completion generally interfere with readability. Maybe if we added a TaskFailedException, so the method would either give a result, or throw one of four exceptions (InterruptedException, TimeOutException, TaskCancelledException and TaskFailedException)?

File: bolts-tasks/src/main/java/bolts/Task.java [hunk 1]
```diff
@@ -115,6 +116,30 @@ public TResult getResult() {
...
+   */
+  public TResult waitAndGetResult(long duration, TimeUnit timeUnit) throws InterruptedException, TimeoutException {
+    synchronized (lock) {
+      if (!waitForCompletion(duration, timeUnit))
+        throw new TimeoutException();
+      if (isFaulted()) {
+        throw new RuntimeException(getError());
+      }
+      if (isCancelled()) {
+        throw new TaskCanceledException(this);
```
Comment (JosephEarl) [inline bolts-tasks/src/main/java/bolts/Task.java]: Use java.util.concurrent.CancellationException to be consistent with other cancellation errors

Comment (JosephEarl) [inline bolts-tasks/src/main/java/bolts/Task.java]: Might be better to throw a java.util.concurrent.CancellationException to be consistent with the other cancellation exceptions

**Comment by stefanrusek** [comment]
@grantland ping

**Comment by grantland** [comment]
Although I'm ashamed that our API didn't match TPL in this case, I'm not entirely sure the benefit of this since it's basically a shortcut for `waitForCompletion() -> getResult()` and the API `waitAndGetResult()` is a bit messy. This being the case I don't think we'll be merging this in.

I will, however, be open to it if others in the community wish for it to be included in Bolts.

**Comment by stefanrusek** [comment]
I don't disagree with you on the cumbersomeness of the API. It makes me sad and highlights the dumbness of checked exceptions, it could trivially be made much less messy, but I doubt the community would go for it.

**Pull request closed without merging**